mini-projects/01-device-driver/laptop-interface/src/lib_serial_ffi.rs

```diff
@@ -0,0 +1,322 @@
+#![allow(non_camel_case_types, dead_code)]
+use libc::size_t;
+use log::{debug, error, info, trace, warn};
+use std::ffi::{CStr, CString};
+use std::os::raw::{c_char, c_int, c_uint, c_void};
+use std::ptr;
+
+use crate::error::SerialError;
+
+#[repr(C)]
+struct sp_port {
+    _private: [u8; 0], // Opaque struct
+}
+
+pub struct SerialPort {
+    sp_port: *mut sp_port,
+}
+
+impl SerialPort {
+    pub fn new(name: CString) -> Result<Self, SerialError> {
+        debug!("Creating new SerialPort with name: {:?}", name);
+        let mut port: *mut sp_port = ptr::null_mut();
+        let result: c_int;
+        unsafe {
+            result = sp_get_port_by_name(name.as_ptr(), &mut port);
+        }
+        trace!(
+            "sp_get_port_by_name returned: {}, port pointer: {:?}",
+            result,
+            port
+        );
+        if port.is_null() {
+            error!("Failed to create port: null pointer");
+            return Err(SerialError::AllocFail);
+        }
+        match result {
+            0 => {
+                info!("Successfully created SerialPort");
+                Ok(SerialPort { sp_port: port })
+            }
+            -1 => {
+                error!("Failed to create port: invalid argument");
+                Err(SerialError::InvalidArgument)
+            }
+            -2 => {
+                error!("Failed to create port: operation failed");
+                Err(SerialError::OperationFailed)
+            }
+            -3 => {
+                error!("Failed to create port: memory allocation error");
+                Err(SerialError::AllocFail)
+            }
+            -4 => {
+                error!("Failed to create port: operation not supported");
+                Err(SerialError::OperationNotSupported)
+            }
+            _ => {
+                error!("Failed to create port: unknown error");
+                Err(SerialError::OperationFailed)
+            }
+        }
```

Comment:
nit: you already do this mapping in `error.rs` - might've been cleaner to leverage that